libbpf-rs/src/object.rs

```diff
@@ -16,108 +18,93 @@ impl Object {
     }
 
     /// Override the generated name that would have been inferred from the constructor.
-    pub fn name<T: AsRef<str>>(&mut self, _name: T) -> &mut Self {
+    pub fn set_name<T: AsRef<str>>(&mut self, _name: T) -> &mut Self {
         unimplemented!();
     }
 
     /// Option to parse map definitions non-strictly, allowing extra attributes/data
-    pub fn relaxed_maps(&mut self) -> &mut Self {
+    pub fn set_relaxed_maps(&mut self, _relaxed_maps: bool) -> &mut Self {
         unimplemented!();
     }
 
-    pub fn get_name(&self) -> &str {
+    pub fn name(&self) -> &str {
         unimplemented!();
     }
 
-    pub fn get_map<T: AsRef<str>>(&mut self, _name: T) -> Option<&mut Map> {
+    pub fn map<T: AsRef<str>>(&mut self, _name: T) -> Option<&mut MapBuilder> {
         unimplemented!();
     }
 
-    pub fn get_prog<T: AsRef<str>>(&mut self, _name: T) -> Option<&mut Program> {
+    pub fn prog<T: AsRef<str>>(&mut self, _name: T) -> Option<&mut ProgramBuilder> {
         unimplemented!();
     }
 }
 
 /// Represents a parsed but not yet loaded map.
-pub struct Map {}
-
-impl Map {
-    pub fn name<T: AsRef<str>>(&mut self, _name: T) -> &mut Self {
-        unimplemented!();
-    }
-
-    pub fn map_type(&mut self, _map_type: MapType) -> &mut Self {
-        unimplemented!();
-    }
-
-    pub fn key_size(&mut self, _size: u32) -> &mut Self {
-        unimplemented!();
-    }
-
-    pub fn value_size(&mut self, _size: u32) -> &mut Self {
-        unimplemented!();
-    }
-
-    pub fn max_entries(&mut self, _entries: u32) -> &mut Self {
-        unimplemented!();
-    }
-
-    pub fn numa_node(&mut self, _node: u32) -> &mut Self {
-        unimplemented!();
-    }
-
-    pub fn btf_fd(&mut self, _fd: u32) -> &mut Self {
-        unimplemented!();
-    }
+///
+/// Some methods require working with raw bytes. You may find libraries such as
+/// [`plain`](https://crates.io/crates/plain) helpful.
+pub struct MapBuilder {}
 
-    pub fn btf_key_type_id(&mut self, _id: u32) -> &mut Self {
+impl MapBuilder {
+    pub fn set_map_ifindex(&mut self, _idx: u32) -> &mut Self {
         unimplemented!();
     }
 
-    pub fn btf_value_type_id(&mut self, _id: u32) -> &mut Self {
+    pub fn set_max_entries(&mut self, _entries: u32) -> &mut Self {
         unimplemented!();
     }
 
-    pub fn map_ifindex(&mut self, _idx: u32) -> &mut Self {
+    pub fn set_initial_value(&mut self, _data: &[u8]) -> &mut Self {
```

Comment:
this can fail for many reasons: not supported map type, invalid data size, etc

Reply:
Hmm but this is a builder method. Wouldn't it be more consistent for the interface to fail in `load()` later?

Reply:
Ah, yeah, I guess for builder it's ok. You'd need to copy _data internally, though, right? Kind of sucks, but probably not a big deal. Having load() failing at the end will make it a bit harder for user to understand what specifically failed, but I think it's fine. Let's make only load() failing.